src/components/Checkins/GoogleMap.tsx

```diff
@@ -0,0 +1,228 @@
+import React, { Component } from 'react';
+import styled from 'styled-components';
+import { makeTimeString } from '../../business/time';
+import Checkin from '../../models/Checkin';
+import CheckinDetails from '../../models/CheckinDetails';
+
+const unselectedMarkerOpacity = 0.5;
+
+const Map = styled.div`
+    height: 100%;
+`;
+
+interface GoogleMapProps {
+    checkins: Checkin[];
+    setSelected(details: CheckinDetails): void;
+}
+
+interface GoogleMapState {
+    hasGoogleLoaded: boolean;
+    mapRef: React.RefObject<HTMLDivElement>;
+}
+
+type PlacesService = google.maps.places.PlacesService;
+type PlaceResult = google.maps.places.PlaceResult;
+type Marker = google.maps.Marker;
+
+class GoogleMap extends Component<GoogleMapProps, GoogleMapState> {
+    private map?: google.maps.Map = undefined;
+    private markers: Marker[] = [];
+    private hasLoadedMarkers = false;
+
+    constructor(props: GoogleMapProps) {
+        super(props);
+        window.googleInit = this.googleInit;
+        this.state = {
+            hasGoogleLoaded: window.hasGoogleLoaded,
+            mapRef: React.createRef<HTMLDivElement>(),
+        };
+    }
+
+    public render = () => {
+        return <Map ref={this.state.mapRef}>Loading map...</Map>;
+    }
+
+    public componentDidUpdate = () => {
+        this.createMapIfNecessary();
+        this.updateMapIfNecessary();
+    }
+
+    public componentDidMount = () => {
+        this.createMapIfNecessary();
+        this.updateMapIfNecessary();
+    }
+
+    // MARK: - Private
+
+    private googleInit = () => {
+        console.log('google init in component');
+        window.hasGoogleLoaded = true;
+        this.setState({ hasGoogleLoaded: true });
+    }
+
+    private createMapIfNecessary = () => {
+        // bounce if google is not initialized
```

Comment:
I think you might be able to handle the bounce here in a [shouldComponentUpdate](https://reactjs.org/docs/react-component.html#shouldcomponentupdate). If this bounce was removed and it instead became `createMap`, you could do (something like):

```typescript
shouldComponentUpdate(_, nextState) {
  if (!this.state.hasGoogleLoaded && nextState.hasGoogleLoaded) {
    this.createMap();
    this.updateMap(); // or whatever
    return true;
  }

  return false;
}